src/trainer.py

```diff
@@ -1155,149 +1117,70 @@ def _save_checkpoint(self, training_state, phase="pretrain", new_best_macro=Fals
-                )
-            )
-            for x in candidate_files:
-                epoch = int(
-                    x.split("model_state_{}_epoch_".format(current_search_phase))[-1].split(".")[0]
-                )
-                if epoch >= max_epoch:
-                    max_epoch = epoch
-                    to_return = x
-            return to_return.split("model_state_")[-1]
-
-    def _restore_checkpoint(self, search_phases_in_priority_order=["pretrain"]):
+    def _restore_checkpoint(self, phase, tasks=None):
         """
         Restores a model from a serialization_dir to the last saved checkpoint.
         This includes an epoch count and optimizer state, which is serialized separately
-        from  model parameters. This function should only be used to continue training -
+        from model parameters. This function should only be used to continue training -
         if you wish to load a model for inference/load parts of a model into a new
         computation graph, you should use the native Pytorch functions:
         `` model.load_state_dict(torch.load("/path/to/model/weights.th"))``
 
+        We restore based on the phase. If phase=target_train, we start from the last 
+        target task and work backwards, to find the most recent checkpoint in the target 
+        train phase. If phase=pretrain, we check for checkpoints in the main run 
+        directory.
         Returns
         -------
-        epoch
-            The epoch at which to resume training.
+        epoch: The epoch at which to resume training.
         """
-
-        suffix_to_load = self._find_last_checkpoint_suffix(
-            search_phases_in_priority_order=search_phases_in_priority_order
+        suffix_to_load = None
+        task_directory, epoch, suffix = check_for_previous_checkpoints(
+            self._serialization_dir, tasks, phase, load_model=True
         )
-        assert suffix_to_load, "No checkpoint found."
-        log.info("Found checkpoint {}. Loading.".format(suffix_to_load))
-
-        model_path = os.path.join(self._serialization_dir, "model_state_{}".format(suffix_to_load))
-        training_state_path = os.path.join(
-            self._serialization_dir, "training_state_{}".format(suffix_to_load)
-        )
-        task_state_path = os.path.join(
-            self._serialization_dir, "task_state_{}".format(suffix_to_load)
-        )
-        metric_state_path = os.path.join(
-            self._serialization_dir, "metric_state_{}".format(suffix_to_load)
+        assert epoch > -1, "No checkpoint found."
+        log.info("Found checkpoint {}. Loading.".format(suffix))
+        if task_directory is None:
+            task_directory = ""
+        model_path = os.path.join(
+            self._serialization_dir, task_directory, "_".join(["model", suffix])
         )
+        training_state_path = model_path.replace("model", "training")
```

Comment:
Just noticed a major risk here: If someone's run_name or exp_name contains the word 'model' (I could easily imagine a exp name called something like 'sweep_model_size'), that'll get replaced, too. Instead, we should just templatically generate all four names.